include/irr/asset/IAsset.h

```diff
@@ -167,25 +173,81 @@ class IAsset : virtual public core::IReferenceCounted
 
         virtual core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const = 0;
 
+		bool restoreFromDummy(IAsset* _other, uint32_t _levelsBelow)
+		{
+			assert(getAssetType() == _other->getAssetType());
+
+			if (!canBeRestoredFrom_recurseDAG(_other))
+				return false;
+
+			restoreFromDummy_impl(_other, _levelsBelow);
+			return true;
+		}
+
+		bool canBeRestoredFrom(const IAsset* _other) const
```

Comment:
ok lets take it from the top.

`canBeRestoredFrom(const IAsset* _other)` as you have it now, should be called `canBeRestoredFrom(const IAsset* _other, uint32_t levelsBelow=~0u)`

Because one SHOULD call `canBeRestoredFrom` with the exact same parameters before calling `restoreFromDummy` via the asset manager, to make sure we don't fuck up during the restore.

so a call to `canBeRestoredFrom_recurseDAG` is in order at the end

ALSO THE CURRENT CODE OF THIS FUNCTION SHOULD MOVE TO A FUNCTION CALLED `canBeRestoredFrom_self`

Reply:
The general idea

User calls `restoreFromDummy` via Asset Manager, and is supposed to check if calling `restoreFromDummy` would result in catastrophic failure by first making sure calling `canBeRestoredFrom` with identical arguments returns true.

Therefore, `canBeRestoredFrom` needs to only check for "asset matching", NOT mutability and dummy states of particular assets.

I propose a `willSelfBeRestoredFrom` which would do the current checks that your current version of `canBeRestoredFrom` does, and you'd use that for the `const bool restorable` in your implementations of restoration (the non recusive portion if-statement).